Allow opening a different project

Cuperino opened the pull request:
Opening a new project closes all files and replaces the current project.

narnaud reviewed (commented):
I see multiple issues here.

On the form:
- there are 2 commits together, one is the close tab that I don't want
- you are not using conventional commit

On the code itself, changing the `Project` root multiple times is a bit more involved, as project settings are loaded in `Project::setRoot`.
Also it will require to stop and restart the LSP clients...

I'm not sure it's super important to be able to open multiple projects. If you want to continue in this direction, go ahead, but there's more work than that.
Otherwise close the PR.

Cuperino commented:
> there are 2 commits together, one is the close tab that I don't want

This was intentional because the changes in `feat: allow opening a different project` depended on the changes done to the `closeDocument` function in `feat: make document tabs easily closable via mouse`. I didn't plan  on merging this PR until the one with those changes had been merged.

What I've done here now is duplicate the code in each PR, so they can be merged individually in any order. Whichever PR is to be merged after will have to be re-based and have its conflicts resolved. This isn't a big deal, but it's what I was trying to avoid.

>  you are not using conventional commit
Fixed.

> On the code itself, changing the Project root multiple times is a bit more involved, as project settings are loaded in Project::setRoot.
Fixed by saving settings if settings would've been saved when closing the app, and then reloading all settings in order so everything's anew.

> Also it will require to stop and restart the LSP clients...

Done. Tho I'm not sure of how to properly test this...

Cuperino commented on src/core/settings.cpp:
```diff
@@ -85,6 +90,9 @@ void Settings::loadUserSettings()
 
 void Settings::loadProjectSettings(const QString &rootDir)
 {
+    saveOnExit();
```
`saveSettings` already existed, so I went with `saveIfApplicable` instead.

Cuperino commented on src/core/settings.cpp:
```diff
@@ -63,6 +61,13 @@ Settings::~Settings()
 }
 
+void Settings::loadBaseSettings()
+{
```
`loadBaseSettings` calls `loadKnutSettings`, which cleanly initializes m_settings by reading, parsing and assigning the contents of ":/core/settings.json". That path is for a Qt Resource, meaning we can guarantee that it'll always be present.

Instead of redundantly clearing the contents of m_settings, what I've done is add an error log and a Q_UNREACHABLE statement after ":/core/settings.json" is read, so Knut stops in the event of a regression in which the resource file goes missing.
